crates/uv-publish/src/lib.rs

```diff
@@ -0,0 +1,784 @@
+    /// otherwise incorrect.\n\n\nError: Use 'source' as Python version for an sdist.\n\n",
+    /// "code": "400 Error: Use 'source' as Python version for an sdist.",
+    /// "title": "Bad Request"}
+    /// ```
+    ///
+    /// ```json
+    /// {"message": "Access was denied to this resource.\n\n\nInvalid or non-existent authentication
+    /// information. See https://test.pypi.org/help/#invalid-auth for more information.\n\n",
+    /// "code": "403 Invalid or non-existent authentication information. See
+    /// https://test.pypi.org/help/#invalid-auth for more information.",
+    /// "title": "Forbidden"}
+    /// ```
+    /// ```json
+    /// {"message": "Access was denied to this resource.\n\n\n\n\n",
+    /// "code": "403 Username/Password authentication is no longer supported. Migrate to API
+    /// Tokens or Trusted Publishers instead. See https://test.pypi.org/help/#apitoken and
+    /// https://test.pypi.org/help/#trusted-publishers",
+    /// "title": "Forbidden"}
+    /// ```
+    ///
+    /// For context, for the last case twine shows:
+    /// ```text
+    /// WARNING  Error during upload. Retry with the --verbose option for more details.
+    /// ERROR    HTTPError: 403 Forbidden from https://test.pypi.org/legacy/
+    ///          Username/Password authentication is no longer supported. Migrate to API
+    ///          Tokens or Trusted Publishers instead. See
+    ///          https://test.pypi.org/help/#apitoken and
+    ///          https://test.pypi.org/help/#trusted-publishers
+    /// ```
+    ///
+    /// ```text
+    /// INFO     Response from https://test.pypi.org/legacy/:
+    ///          403 Username/Password authentication is no longer supported. Migrate to
+    ///          API Tokens or Trusted Publishers instead. See
+    ///          https://test.pypi.org/help/#apitoken and
+    ///          https://test.pypi.org/help/#trusted-publishers
+    /// INFO     <html>
+    ///           <head>
+    ///            <title>403 Username/Password authentication is no longer supported.
+    ///          Migrate to API Tokens or Trusted Publishers instead. See
+    ///          https://test.pypi.org/help/#apitoken and
+    ///          https://test.pypi.org/help/#trusted-publishers</title>
+    ///           </head>
+    ///          <body>
+    ///           <h1>403 Username/Password authentication is no longer supported.
+    ///         Migrate to API Tokens or Trusted Publishers instead. See
+    ///          https://test.pypi.org/help/#apitoken and
+    ///          https://test.pypi.org/help/#trusted-publishers</h1>
+    ///            Access was denied to this resource.<br/><br/>
+    /// ```
+    ///
+    /// In comparison, we now show (line-wrapped for readability):
+    ///
+    /// ```text
+    /// error: Failed to publish `dist/astral_test_1-0.1.0-py3-none-any.whl` to `https://test.pypi.org/legacy/`
+    ///   Caused by: Incorrect credentials (status code 403 Forbidden): 403 Username/Password
```

Comment:
Why do we say "Incorrect credentials" here? That seems misleading and it's not obvious where it comes from given all the above.

Reply:
xref https://github.com/astral-sh/uv/pull/7475/files#r1771980162

Reply:
tbh i don't have any strong feelings about what we say here, we can't tell what the real reason is (was your password wrong? are you lacking permission for the project? did you make a typo in your trusted publishing configuration?), and the interesting part of the error is the message from the server.

Reply:
I'd prefer the more general message so we're not misleading the user